code/parachain/frame/staking-rewards/src/lib.rs

```diff
@@ -1664,85 +1663,101 @@ fn update_rewards_pool<T: Config>(
-		);
-
-		if !newly_accumulated_rewards.is_zero() {
-			T::Assets::release(
-				asset_id,
-				pool_account,
-				newly_accumulated_rewards.into(),
-				false, // not best effort, entire amount must be released
-			)
-			.expect("funds should be available to release based on previous check; qed;");
+	// if no periods have been surpassed, short-circuit
+	let Some(periods_surpassed) = NonZeroU64::new(periods_surpassed) else {
+		return Ok(())
+	};
 
-			reward.total_rewards = new_total_rewards.into();
-			reward.last_updated_timestamp = last_updated_timestamp;
+	let total_locked_rewards: u128 = T::Assets::balance_on_hold(asset_id, pool_account).into();
+
+	// the maximum amount repayable given the reward rate.
+	// i.e. if total locked is 50, and the reward rate is 15, then this would be 3
+	let maximum_releasable_periods =
+		NonZeroU128::new(<u128 as Div<NonZeroU128>>::div(total_locked_rewards, reward_rate_amount))
+			// if the maximum releasable periods is zero, then that means the pot doesn't have
+			// enough in it to fund a single period.
+			.ok_or(RewardAccumulationCalculationError::RewardsPotEmpty)?;
+
+	let releasable_periods_surpassed =
+		cmp::min(maximum_releasable_periods, periods_surpassed.into());
+
+	// SAFETY: Usage of mul is safe here. See the following proof:
+	//
+	// ```plaintext
+	//                                    (  total_locked_rewards           elapsed_time        )
+	// releasable_periods_surpassed = min ( ---------------------- , -------------------------- )
+	//                                    (   reward_rate_amount     reward_rate_period_seconds )
+	// 											   ^ LHS				       ^ RHS
+	//
+	//                             {                      (  total_locked_rewards  )
+	//                             { reward_rate_amount * ( ---------------------- ),     if LHS <= RHS
+	//                             {                      (   reward_rate_amount   )
+	// newly_accumulated_rewards = {
+	//                             {                      (        elapsed_time        )
+	//                             { reward_rate_amount * ( -------------------------- ), if RHS < LHS
+	//                             {                      ( reward_rate_period_seconds )
+	// ```
+	//
+	// Note that the second function of the piecewise definition above will always be <= the first,
+	// since if it were larger it would not have been selected; and since the first function is
+	// always infallible (the resulting value will always be <= the total_locked_rewards input due
+	// to integer division), the second part is also infallible if it is less than the first.
+	let newly_accumulated_rewards = releasable_periods_surpassed
+		.checked_mul(reward_rate_amount)
+		.expect("should not fail; see above for proof; qed;");
+
+	let new_total_rewards = newly_accumulated_rewards
+		.get()
+		.checked_add(reward.total_rewards.into())
+		// TODO(benluelo): Rename to overflow to be more specific
```

Comment:
?

Reply:
forgot to remove this comment, thanks :+1: